elastic/kibana
# [Response Ops][Actions] Allow executing actions directly via `UnsecuredActionsClient`

## Summary

Adds unsecured version of direct action execution that is exposed via the `unsecuredActionsClient` on the actions plugin start contract. This version doesn't expect any user request and doesn't check RBAC. Saved objects are accessed via the internal saved objects repository (instead of a saved objects client scoped to the request) and the connector token client uses an ES client that uses the internal user (instead of scoped to the request).

Similar to the `bulkExecutionEnqueuer`, this function expects a `requesterId` which is checked against an allow-list. For this PR we added a `background_task` requester since we expect this to be used by background tasks. If the `relatedSavedObjects` array is provided, we inject  `BACKGROUND_TASK` source data that will be persisted in the event log doc for this execution that aids in traceability and debugging.

## To Verify
1. Run this PR and create a connector via the UI or API. (I created a Slack API connector). Grab the ID of this connector.
2. In `x-pack/plugins/alerting/server/plugin.ts` in the start function, add the following:

```
const unsecuredActionsClient = plugins.actions.getUnsecuredActionsClient();
unsecuredActionsClient.execute({
  requesterId: 'background_task',
  id: <connectorId>,
  spaceId: 'default',
  // params format depends on what connector type you created
  params: {
    subAction: 'postMessage,
    subActionParams: {
      channelIds: [<channelId>],
      text: 'yoyoyo'
    }
  },
  relatedSavedObjects: [{
    type: 'task',
    // this should be the ID of the task instance calling the execute function
    id: 'abc',
    // this should be the task type of the task instance calling the execute function
    typeId: 'taskType'
  }]
});
```
3. Restart Kibana and verify that the action was executed and an event log doc was indexed.
